hbt/production/tau.py

```diff
@@ -0,0 +1,317 @@
+    correction_set = correctionlib.CorrectionSet.from_string(
+        bundle.files.tau_sf.load(formatter="gzip").decode("utf-8"),
+    )
+    self.id_vs_jet_corrector = correction_set[f"{self.config_inst.x.tau_tagger}VSjet"]
+    self.id_vs_e_corrector = correction_set[f"{self.config_inst.x.tau_tagger}VSe"]
+    self.id_vs_mu_corrector = correction_set[f"{self.config_inst.x.tau_tagger}VSmu"]
+
+
+@producer(
+    uses={
+        "channel_id", "single_triggered", "cross_triggered",
+        "nTau", "Tau.pt", "Tau.decayMode",
+    },
+    produces={
+        "tau_trigger_weight",
+        "tau_trigger_weight_tautau_up", "tau_trigger_weight_tautau_down",
+        "tau_trigger_weight_etau_up", "tau_trigger_weight_etau_down",
+        "tau_trigger_weight_mutau_up", "tau_trigger_weight_mutau_down",
+    },
+)
+def trigger_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Producer for trigger scale factors derived by the TAU POG. Requires an external file in the
+    config as (e.g.):
+
+    .. code-block:: python
+
+        "tau_sf": ("/afs/cern.ch/user/m/mrieger/public/mirrors/jsonpog-integration-f018adfb/POG/TAU/2017_UL/tau.json.gz", "v1"),  # noqa
+
+    which should be a correctionlib-style json file from which a correction set named "tau_trigger"
+    is extracted.
+
+    Resources:
+    https://twiki.cern.ch/twiki/bin/view/CMS/TauIDRecommendationForRun2?rev=113
+    https://gitlab.cern.ch/cms-nanoAOD/jsonpog-integration/-/blob/849c6a6efef907f4033715d52290d1a661b7e8f9/POG/TAU
+    """
+    # fail when running on data
+    if self.dataset_inst.is_data:
+        raise ValueError("attempt to compute trigger weights in data")
+
+    # get channels from the config
+    ch_etau = self.config_inst.get_channel("etau")
+    ch_mutau = self.config_inst.get_channel("mutau")
+    ch_tautau = self.config_inst.get_channel("tautau")
+
+    # helper to bring a flat sf array into the shape of taus, and multiply across the tau axis
+    reduce_mul = lambda sf: ak.prod(layout_ak_array(sf, events.Tau.pt), axis=1, mask_identity=False)
+
+    # the correction tool only supports flat arrays, so convert inputs to flat np view first
+    pt = flat_np_view(events.Tau.pt, axis=1)
+    dm = flat_np_view(events.Tau.decayMode, axis=1)
+
+    #
+    # compute nominal trigger weight
+    #
+
+    # define channel / trigger dependent masks
+    channel_id = events.channel_id
+    single_triggered = events.single_triggered
+    dm_mask = (
+        (events.Tau.decayMode == 0) |
+        (events.Tau.decayMode == 1) |
+        (events.Tau.decayMode == 10) |
+        (events.Tau.decayMode == 11)
+    )
+    tautau_mask = flat_np_view(
+        dm_mask & (channel_id == ch_tautau.id),
+        axis=1,
+    )
+    # not existing yet
+    # tautauvbf_mask = flat_np_view(dm_mask & (channel_id == ch_tautau.id), axis=1)
+    etau_mask = flat_np_view(
+        dm_mask & (channel_id == ch_etau.id) & single_triggered & (events.Tau.pt >= 25.0),
```

Comment:
The pT cuts for the single tau channels depend on the current recommendations. Wouldn't it be easier to configer this via the config?

Reply:
The cut here comes from a limitation of the json file being read by the correctionlib. The tau pog does not provide SFs for pt < 25, which makes the number appearing in the producer here indeed somewhat awkward :/

In the selection, we have different pt thresholds depending on the channel and the fired triggers.

I think the optimal solution would be that the correctionlib would just return `1` if pt < 25, but currently an exception is raised. I think this will be fixed in the future in the json file itself, so I thought we could leave that here. Would you be ok with that?

Reply:
Sure :+1: Putting such a cut into a config might also be very awkward in any case, so let's hope that this is resolved centrally in teh correctionlib